### DISCUSSION TITLE
How to show a default value in a comboBox for a modal window with "setSelectedItemIndex"

### DISCUSSION THREAD

**Discussion by damiensellier** [RomanKubiak/ctrlr#246] [Q&A] [answered]
Hi,

I made a modalBox to select the input and output midi devices.
Everything is fine but I want to select/show the combobox to the current value as default value of the dropdown but I cannot achieve it. Now it's the first one of the array.

Here is my code  :

```
function MidiDeviceSelect()

console("Current Input device : "..L(panel:getPropertyString("panelMidiInputDevice")))
console("Current Controller device : "..L(panel:getPropertyString("panelMidiControllerDevice")))
console("Current Output device : "..L(panel:getPropertyString("panelMidiOutputDevice")))

	saMidiInputDevices = StringArray()
	for i=0,(utils.getMidiInputDevices():size()) do
		saMidiInputDevices:set(i, utils.getMidiInputDevices():get(i))
	end

	saMidiOutputDevices = StringArray()
	for i=0,(utils.getMidiOutputDevices():size()) do
		saMidiOutputDevices:set(i, utils.getMidiOutputDevices():get(i))
	end


	local CurrentMidiInputDevice = panel:getProperty("panelMidiInputDevice")
	local CurrentMidiControllerDevice = panel:getProperty("panelMidiControllerDevice")
	local CurrentMidiOutputDevice = panel:getProperty("panelMidiOutputDevice")


	modalWindow:getComboBoxComponent("cmbMidiOutputDevices"):setSelectedItemIndex(saMidiInputDevices, CurrentMidiInputDevice) -- Set Default Value to current midi input
	modalWindow:getComboBoxComponent("cmbMidiInputDevices"):setSelectedItemIndex(saMidiOutputDevices, CurrentMidiOutputDevice) -- Set Default Value to current midi output


	modalWindow = AlertWindow("MIDI I/O Devices", "Select your input and output devices. \n Never set input and output the same to avoid loops.", AlertWindow.NoIcon)
	modalWindow:addButton("    OK    ", 1, KeyPress(KeyPress.returnKey),KeyPress())
	modalWindow:addButton("Cancel", 0, KeyPress(KeyPress.escapeKey),KeyPress())


	modalWindow:addComboBox ("cmbMidiInputDevices", saMidiInputDevices, "MIDI Input Device [Virtual Port MCU IN]")

	--modalWindow:addComboBox ("cmbMidiControllerDevices", saMidiOutputDevices, "MIDI Controller Device")

	modalWindow:addComboBox ("cmbMidiOutputDevices", saMidiOutputDevices, "MIDI Output Device [Midi Fighter Twister]")


	ret = modalWindow:runModalLoop()


	if ret == 1 then
		panel:setPropertyString("panelMidiInputDevice",modalWindow:getComboBoxComponent("cmbMidiInputDevices"):getText())
		--panel:setPropertyString("panelMidiControllerDevice",modalWindow:getComboBoxComponent("cmbMidiControllerDevices"):getText())
		panel:setPropertyString("panelMidiOutputDevice",modalWindow:getComboBoxComponent("cmbMidiOutputDevices"):getText())

	end

	modalWindow:setVisible (false)

end -- end f
```

The part that is problematic is this one : 

```
	local CurrentMidiInputDevice = panel:getProperty("panelMidiInputDevice")
	local CurrentMidiOutputDevice = panel:getProperty("panelMidiOutputDevice")

	modalWindow:getComboBoxComponent("cmbMidiOutputDevices"):setSelectedItemIndex(saMidiInputDevices, CurrentMidiInputDevice) -- Set Default Value to current midi input
	modalWindow:getComboBoxComponent("cmbMidiInputDevices"):setSelectedItemIndex(saMidiOutputDevices, CurrentMidiOutputDevice) -- Set Default Value to current midi output
```

I am sure the arguments are incorrect for **setSelectedItemIndex()**

Can you please help with that?

Thanks a lot in advance, have a nice weekend

Damien

**Comment by Ted-juh** [comment]
void setSelectedItemIndex(const int index, const bool dontNotify);

So it seems it takes a number for the index and a true/ false for the bool.

setSelectedItemIndex(0, true)

**Comment by dnaldoog** [comment, marked as answer]
This code works for me. Looking through it, I can't explain why. Something to do with _NotificationType_ as Tedjuh said?
 
<pre>
<code>
midi_in_selection = function(--[[ CtrlrComponent --]] comp --[[ MouseEvent --]], event)
    local myChannelIn = tonumber(panel:getProperty("panelMidiInputChannelDevice"))
    local myInputDevice = panel:getProperty("panelMidiInputDevice")
    local deviceId = 0

    saMidiInputDevices = StringArray()
    for i = 0, (utils.getMidiInputDevices():size()) do
        saMidiInputDevices:set(i, utils.getMidiInputDevices():get(i))
        if utils.getMidiInputDevices():get(i) == myInputDevice then
            deviceId = i
        end
    end

    saMidiInputChannel = StringArray()
    saMidiInputChannel:set(0, "All channels")
    for i = 1, 16 do
        saMidiInputChannel:set(i, tostring(i))
    end

    modalWindow = AlertWindow("\nMIDI INPUT", "", AlertWindow.NoIcon)
    modalWindow:addButton("    OK    ", 1, KeyPress(KeyPress.returnKey), KeyPress())
    modalWindow:addButton("Cancel", 0, KeyPress(KeyPress.escapeKey), KeyPress())
    modalWindow:addComboBox("cmbMidiInputDevices", saMidiInputDevices, "Device")
    modalWindow:addComboBox("cmbMidiInputChannel", saMidiInputChannel, "MIDI Input Channel")
    modalWindow:addTextBlock("")

    for i = 0, (utils.getMidiInputDevices():size()) do
    end

    modalWindow:getComboBoxComponent("cmbMidiInputDevices"):setSelectedItemIndex(deviceId, 2)
    modalWindow:getComboBoxComponent("cmbMidiInputChannel"):setSelectedItemIndex(myChannelIn, 2)

    ret = modalWindow:runModalLoop()

    if ret == 1 then
        panel:setPropertyString(
            "panelMidiInputDevice",
            modalWindow:getComboBoxComponent("cmbMidiInputDevices"):getText()
        )
        panel:setPropertyInt(
            "panelMidiInputChannelDevice",
            modalWindow:getComboBoxComponent("cmbMidiInputChannel"):getSelectedItemIndex()
        )
    end

    modalWindow:setVisible(false)
end

</code>
</pre>

**Reply by damiensellier** [reply]
@dnaldoog Thanks! It's exactly what I wanted. 

Is there a function like **utils.getMidiInputDevices** specific for **Controller** Input Devices ?
**utils.getMidiControllerDevices** doesn't work. 
And all the script I checked on ctrlr.org have the **utils.getMidiInputDevices** even for the controller.

Thanks a lot!
